File: Leetcode(python)/002_AddTwoNumbers.py
# coding:utf-8
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        addends = l1, l2 #组成个列表
        dummy = end = ListNode(0)
        carry = 0
        while addends or carry:
            carry += sum(a.val for a in addends)
            addends = [a.next for a in addends if a.next]
            end.next = end = ListNode(carry % 10)
            carry /= 10
        return dummy.next

File: Leetcode(Go)/002_AddTwoNumbers.go
package Leetcode


type ListNode struct {
	Val int
    Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    dummy := &ListNode{ Val: 0 }//构造节点
    p := dummy
    carry := 0

    for l1 != nil || l2 != nil {//for可以当while用
        add := carry
        if l1 != nil {
            add += l1.Val
            l1 = l1.Next
        }
        if l2 != nil {
            add += l2.Val
            l2 = l2.Next
        }
        carry = add / 10
        add = add % 10
        p.Next = &ListNode{ Val: add }
        p = p.Next
    }

    if carry != 0 {
        p.Next = &ListNode{ Val: carry }
    }

    return dummy.Next
}